Reduce unsafeness in DeviceController classes

annevk opened the pull request:
#### 8e11110eadb783c1d43fee94e4e3997c8ebf053c
<pre>
Reduce unsafeness in DeviceController classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=301997">https://bugs.webkit.org/show_bug.cgi?id=301997</a>

Reviewed by Chris Dumez.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

Canonical link: <a href="https://commits.webkit.org/302584@main">https://commits.webkit.org/302584@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dfda42fcd40331c43ceb027b2476c4b6686676bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1844 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136972 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81028 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9714281e-ebb9-4f4f-97ac-251a2ba4fdf5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131458 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1720 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98711 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66563 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/76609ad6-155b-4496-97e6-a38f96cb7f00) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132534 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1375 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116068 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79374 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e7b7dca8-29a9-47e6-a78d-08904e7876e9) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1291 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34199 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80247 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109767 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34697 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139452 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1634 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1558 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107231 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1676 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112409 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107076 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27262 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1334 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30921 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54353 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1705 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65068 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1525 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1559 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1627 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/mock/DeviceOrientationClientMock.cpp:
```diff
@@ -34,7 +34,7 @@ namespace WebCore {
 
 DeviceOrientationClientMock::DeviceOrientationClientMock()
-    : m_controller(0)
+    : m_controller(nullptr)
```
This line is no longer needed, m_controller now gets default-initialized to nullptr already now that you made it a smart pointer.

cdumez commented on Source/WebCore/platform/mock/DeviceOrientationClientMock.h:
```diff
@@ -23,10 +23,10 @@
+#pragma once
 
 #include <WebCore/DeviceOrientationClient.h>
+#include <WebCore/DeviceOrientationController.h>
```
Instead of doing this and making the header private, can we keep forward-declaring and simply add a class destructor that is out of line?

annevk replied:
Is there an example of that somewhere? Not entirely sure what that would look like.

cdumez replied:
Essentially drop this new include, bring back the forward declaration you had dropped. Then add this to the class right after the constructor:
```
~DeviceOrientationClientMock();
```

Then add this to the `.cpp`:
```
DeviceOrientationClientMock::~DeviceOrientationClientMock() = default;
```

cdumez replied:
Essentially, if you don't declare the destructor and move its implementation out of line, then the compiler is implicitly generating an inline version for you in the header. And because this destructor destroys the data member that is a `CheckedPtr<DeviceOrientationController>`, it has to know about the type in the header and you need the include. So instead, we move the destructor out of line so only the cpp file needs to know about the type, not the header.

annevk replied:
Thanks!

cdumez reviewed (approved):
LGTM

annevk committed:
Reduce unsafeness in DeviceController classes
https://bugs.webkit.org/show_bug.cgi?id=301997

Reviewed by Chris Dumez.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

Canonical link: https://commits.webkit.org/302584@main
